lib/charliecloud.py

```diff
@@ -597,47 +587,145 @@ def from_tree(self, t):
+         # include putting the grammar into Lark (this can be gotten by
+         # reading the RFCs enough) or using the www-authenticate library [3].
+         #
+         # [1]: https://stackoverflow.com/a/1349528
+         # [2]: https://stackoverflow.com/a/1547940
+         # [3]: https://pypi.org/project/www-authenticate
+         DEBUG("requesting auth parameters")
+         res = self.get_raw(url, expected_statuses=(401,))
+         if ("WWW-Authenticate" not in res.headers):
+            FATAL("WWW-Authenticate header not found")
+         auth = res.headers["WWW-Authenticate"]
+         if (not auth.startswith("Bearer ")):
+            FATAL("authentication scheme is not Bearer")
+         authd = dict(re.findall(r'(?:(\w+)[:=] ?"?([\w.~:/?#@!$&()*+,;=\'\[\]-]+)"?)+', auth))
+         DEBUG("WWW-Authenticate parse: %s" % authd, v=2)
+         for k in ("realm", "service", "scope"):
+            if (k not in authd):
+               FATAL("WWW-Authenticate missing key: %s" % k)
+         # Request auth token.
+         DEBUG("requesting anonymous auth token")
+         res = self.get_raw(authd["realm"], expected_statuses=(200,403),
+                            params={"service": authd["service"],
+                                    "scope": authd["scope"]})
+         if (res.status_code == 403):
+            INFO("anonymous access rejected")
+            username = input("Username: ")
+            password = getpass.getpass("Password: ")
+            auth = requests.auth.HTTPBasicAuth(username, password)
+            res = self.get_raw(authd["realm"], auth=auth,
+                               params={"service": authd["service"],
+                                       "scope": authd["scope"]})
+         token = res.json()["token"]
          DEBUG("got token: %s..." % (token[:32]))
-         s.headers.update({ "Authorization": "Bearer %s" % token })
-         self._session = s
-      return self._session
+         self.auth = self.Bearer_Auth(token)
 
    def close(self):
-      if (self._session is not None):
-         self._session.close()
+      if (self.session is not None):
+         self.session.close()
+
+   def get(self, url, path, headers=dict()):
+      """GET url, passing headers, including authentication and session magic,
+         and write the body of the response to path."""
+      DEBUG("GETting: %s" % url)
+      self.session_init_maybe()
+      self.authenticate_maybe(url)
+      res = self.get_raw(url, headers)
+      try:
+         fp = open(path, "wb")
+         fp.write(res.content)
+         fp.close()
+      except OSError as x:
+         FATAL("can't write: %s: %s" % (path, x))
+
+   def get_layer(self, hash_, path):
+      "GET the layer with hash hash_ and save it at path."
+      # /v1/library/hello-world/blobs/<layer-hash>
+      url = self._url_of("blobs", "sha256:" + hash_)
+      accept = "application/vnd.docker.image.rootfs.diff.tar.gzip"
```

Comment:
If we don't already, we should include a URL where this media type stems from.

Reply:
Turns out we do, on line 208 in `Image.download()`.